rat/fit.py

```diff
@@ -139,18 +157,84 @@ def save(self, folder, overwrite=False):
             f.write(self.__class__.__name__)
 
 
+@dataclass
+class GeneratedQuantitiesFunctionGenerator(RatWalker):
+    variable_table: VariableTable
+    trace_table: TraceTable
+    parameters: Dict[str, Any]
+    traced_nodes: List[ast.Variable] = field(default_factory=list)
+
+    def process_node(self, node: ast.ModelBase):
+        primary_node = get_primary_ast_variable(node)
+        primary_name = primary_node.name
+        primary_variable = self.variable_table[primary_name]
+
+        self.traced_nodes = []
+
+        self.walk(node)
+
+        traced_values = tuple(self.trace_table[traced_node].array for traced_node in self.traced_nodes)
+
+        def mapper(*mapper_arguments):
+            executor = CodeExecutor(dict(zip(self.traced_nodes, mapper_arguments)), self.parameters)
+            return executor.walk(node)
+
+        if primary_variable.argument_count > 0:
+            return jax.vmap(mapper)(*traced_values)
+        else:
+            return mapper(*traced_values)
+
+    def walk_Variable(self, node: ast.Variable):
+        node_variable = self.variable_table[node.name]
+        if node_variable.argument_count > 0:
+            if node in self.trace_table:
+                self.traced_nodes.append(node)
+
+
 class OptimizationFit(Fit):
     """
     Stores optimization results
     """
-
-    def __init__(self, draw_dfs):
+    def __init__(self, model : Model, draw_dfs: Dict[str, pandas.DataFrame], unconstrained_draws: numpy.ndarray):
+        self.model = model
         self.draw_dfs = draw_dfs
+        self.unconstrained_draws = unconstrained_draws
+
+    def expr(self, expression_string : str):
+        # Parse the model to get AST
+        semantics = ModelBuilderSemantics()
+        parser = RatParser(semantics=semantics)
+        # TODO: This lambda is just here to make sure pylance formatting works -- should work
+        # without it as well
+        expression = (lambda: parser.parse(expression_string, start = "expression"))()
+
+        walker = SubscriptTableWalker(self.model.variable_table)
+        walker.process_node(expression)
+        trace_table = walker.trace_table
+
+        @jax.jit
+        @jax.vmap
+        @jax.vmap
+        def run(unconstrained_draws : numpy.ndarray):
```

Comment:
One of the vmaps is for chain and one is for draw

Reply:
ngl that's super confusing syntax D:

Reply:
Yeah, I don't think there's much to be done about the decorators, but in the long run I'd like to make this a regular function and not be inlining it to make the inputs and outputs clear.